packages/kit/src/core/sync/create_manifest_data/index.js

```diff
@@ -229,151 +272,169 @@ export default function create_manifest_data({
 
 	return {
 		assets,
-		layout,
-		error,
 		components,
 		routes,
 		matchers
 	};
 }
 
 /**
- * @param {string} needle
- * @param {string} haystack
+ * @param {string} file
+ * @param {Tree} tree
+ * @param {string[]} extensions
  */
-function count_occurrences(needle, haystack) {
-	let count = 0;
-	for (let i = 0; i < haystack.length; i += 1) {
-		if (haystack[i] === needle) count += 1;
+function trace(file, tree, extensions) {
+	/** @type {Array<string | undefined>} */
+	const layouts = [];
+
+	/** @type {Array<string | undefined>} */
+	const errors = [];
+
+	const parts = file.split('/');
+	const filename = /** @type {string} */ (parts.pop());
+	const extension = /** @type {string} */ (extensions.find((ext) => file.endsWith(ext)));
+	const base = filename.slice(0, -extension.length);
+
+	let layout_id = base.includes('@') ? base.split('@')[1] : DEFAULT;
+
+	// walk up the tree, find which __layout and __error components
+	// apply to this page
+	while (true) {
+		const node = tree.get(parts.join('/'));
+		const layout = node?.layouts[layout_id];
+
+		errors.unshift(node?.error);
+		layouts.unshift(layout?.file);
+
+		if (layout?.name.includes('@')) {
+			layout_id = layout.name.split('@')[1];
+		} else {
+			if (parts.length === 0) break;
```

Comment:
you could change this line together with the code change above
```suggestion
		} else if (parts.length > 0)  {
```

Reply:
unfortunately the algorithm would then be incorrect